frontend/apps/reader/modules/readertoc.lua

```diff
@@ -215,6 +215,27 @@ function ReaderToc:getTocIndexByPage(pn_or_xp)
     return pre_index
 end
 
+function ReaderToc:getAccurateTocIndexByXPointer(xptr)
+    local index = self:getTocIndexByPage(xptr)
+    if self.toc[index + 1] then
+        local next_index_xptr = self.toc[index + 1].xpointer
+        local comparison = self.ui.document:compareXPointers(next_index_xptr, xptr)
+        if comparison and comparison > 0 then
+            return index + 1
+        end
+    end
```

Comment:
I don't think looking at +1 is enough: you can have multiple TOC entries for a single page:

<kbd>![image](https://user-images.githubusercontent.com/24273478/81810090-a3ef0a00-9522-11ea-81a6-6495273246d5.png)</kbd>

These 3 highlights should each have a different chapter (the first one should have the chapter that is shown on a previous page).
Actually, the dogear bookmark says it's in chapter "Etats impériaux" (getTocTitleByPage), which I guess is allright as nobody complains :) but I think it should be too the chapter that started on previous page.